adafruit_httpserver/server.py

```diff
@@ -2,97 +2,220 @@
 #
 # SPDX-License-Identifier: MIT
 """
-`adafruit_httpserver.server.HTTPServer`
+`adafruit_httpserver.server`
 ====================================================
 * Author(s): Dan Halbert, Michał Pokusa
 """
 
 try:
-    from typing import Callable, Protocol, Union
+    from typing import Callable, Protocol, Union, List, Set, Tuple
     from socket import socket
     from socketpool import SocketPool
 except ImportError:
     pass
 
 from errno import EAGAIN, ECONNRESET, ETIMEDOUT
-
-from .exceptions import FileNotExistsError, InvalidPathError
-from .methods import HTTPMethod
-from .request import HTTPRequest
-from .response import HTTPResponse
-from .route import _HTTPRoutes, _HTTPRoute
-from .status import CommonHTTPStatus
-
-
-class HTTPServer:
+from traceback import print_exception
+
+from .authentication import Basic, Bearer, require_authentication
+from .exceptions import (
+    ServerStoppedError,
+    AuthenticationError,
+    FileNotExistsError,
+    InvalidPathError,
+    ServingFilesDisabledError,
+)
+from .methods import GET, HEAD
+from .request import Request
+from .response import Response, FileResponse
+from .route import _Routes, _Route
+from .status import BAD_REQUEST_400, UNAUTHORIZED_401, FORBIDDEN_403, NOT_FOUND_404
+
+
+class Server:
     """A basic socket-based HTTP server."""
 
-    def __init__(self, socket_source: Protocol, root_path: str) -> None:
+    host: str = None
+    port: int = None
+
+    def __init__(
+        self, socket_source: Protocol, root_path: str = None, *, debug: bool = False
+    ) -> None:
         """Create a server, and get it ready to run.
 
         :param socket: An object that is a source of sockets. This could be a `socketpool`
           in CircuitPython or the `socket` module in CPython.
         :param str root_path: Root directory to serve files from
+        :param bool debug: Enables debug messages useful during development
         """
+        self._auths = []
         self._buffer = bytearray(1024)
         self._timeout = 1
-        self.routes = _HTTPRoutes()
+        self._routes = _Routes()
         self._socket_source = socket_source
         self._sock = None
         self.root_path = root_path
+        self.stopped = False
+
+        self.debug = debug
 
-    def route(self, path: str, method: HTTPMethod = HTTPMethod.GET) -> Callable:
+    def route(
+        self,
+        path: str,
+        methods: Union[str, Set[str]] = GET,
```

Comment:
After tinkering a bit more I think this is one cog in the issue about required / default  methods argument value. The way this is typed, and the default value seem to indicate that it accepts either a single method string like `GET` or a set of them like `[GET, POST]`.

But it seems the single method string version does not actually work. 

If you use a route like this
```
@server.route("/hello", GET)
```
It does not get called when user navigates to `http://<server>/hello`.  If you change it to use `[GET]` instead for the argument then it does get called. 

I also verified that changing the default value here to `[GET]` instead of it's current value `GET` does allow you to start using routes with default arguments like:

```
@server.route("/hello")
```
And have it will be called and return it's response.


It seems like at least at one point the intention was to support either single or multiple. I think I'd lean toward just forcing multiple which is how it behaves now if I'm understanding correctly. If user wants to use the single one they can include their single item in the list. By not supporting both types it should make the code implementing the behavior a little easier because it can always just iterate over them and act upon them without needing to worry about which type was sent.

Requiring the list seems to be how Flask works as well if I am understanding the docs at the very bottom of this section: https://flask.palletsprojects.com/en/2.3.x/api/#url-route-registrations inside of the table for the details of the `**options` value

Reply:
As above, this was purely the implementation issue, the intention was to support both single and multiple methods, so typing is ok, now it should be working fine.

Thank you for finding that, it was a very simple mistake that might confuse others, especially considering that there was no exception raised.